### PR TITLE
SONAR-6713 Load rules in global container

### PR DESCRIPTION
**Pull request by dbmeneses**

### CODE REVIEW

File: sonar-batch/src/main/java/org/sonar/batch/rule/DefaultRulesLoader.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+      RulesSearchResult moreRules = RulesSearchResult.fromJson(wsLoader.loadString(getUrl(i)));
+      if (moreRules.getRules().isEmpty()) {
+        break;
+      }
+      rules.getRules().addAll(moreRules.getRules());
+    }
+    return rules;
+  }
+
+  private String getUrl(int page) {
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/rule/DefaultRulesLoader.java]: MAJOR Make "getUrl" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

File: sonar-batch/src/main/java/org/sonar/batch/rule/DefaultRulesLoader.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.batch.rule;
+
+import org.sonar.batch.protocol.input.Rule;
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/rule/DefaultRulesLoader.java]: MINOR Remove this unused import 'org.sonar.batch.protocol.input.Rule'. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-batch/src/main/java/org/sonar/batch/scan/report/IssuesReportBuilder.java [hunk 3]
```diff
@@ -26,8 +29,6 @@
 import org.sonar.core.issue.DefaultIssue;
 import org.sonar.api.resources.Project;
 import org.sonar.api.rule.RuleKey;
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/scan/report/IssuesReportBuilder.java]: MINOR Remove this unused import 'org.sonar.api.rule.RuleKey'. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-batch/src/main/java/org/sonar/batch/cpd/CpdComponents.java [hunk 4]
```diff
@@ -28,7 +30,7 @@
   private CpdComponents() {
   }
 
-  public static List all() {
+  public static List<? extends Object> all() {
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/cpd/CpdComponents.java]: MAJOR Remove usage of generic wildcard type. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1452)

File: sonar-batch/src/main/java/org/sonar/batch/rule/RuleFinderCompatibility.java [hunk 5]
```diff
@@ -96,28 +95,27 @@ public Rule find(RuleQuery query) {
   }
 
   private Collection<Rule> byRepository(RuleQuery query) {
-    return Collections2.transform(activeRules.findByRepository(query.getRepositoryKey()), new Function<ActiveRule, Rule>() {
+    return Collections2.transform(rules.findByRepository(query.getRepositoryKey()), new Function<org.sonar.api.batch.rule.Rule, Rule>() {
```
Comment (SonarTech) [inline sonar-batch/src/main/java/org/sonar/batch/rule/RuleFinderCompatibility.java]: MAJOR Make this a named "static" inner class. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2694)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues:
* CRITICAL 1 critical
* MAJOR 3 major
* MINOR 2 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* CRITICAL [Method 'public org.sonar.api.batch.rule.Rule findByInternalKey(java.lang.String, java.lang.String)' has been added to an interface](https://github.com/SonarSource/sonarqube/blob/7c0f77be916900bfd243183c3062c97ca53818e8/sonar-plugin-api/src/main/java/org/sonar/api/batch/rule/Rules.java) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=clirr%3Aclirr-api-break)

**Comment by henryju** [comment]
Do you plan to remove the "name" and "internalKey" attributes on ActiveRule to avoid duplication with Rule?

**Comment by dbmeneses** [comment]
@henryju org.sonar.api.batch.rule.ActiveRule doesn't have "name". It has internalKey which won't be used (as discussed). It is already marked with @CheckForNull, even though with a different purpose, I think (null if plugin doesn't declare it). I would just modify javadoc.

**Comment by dbmeneses** [comment]
It was merged

**Pull request closed without merging**